### PR TITLE
List of 2Ds -> 3D Tensor, seq2seq_loss

### PR DESCRIPTION
**Pull request by alrojo**
Use full tensor and reshape instead of evaluating at every timestep, computationally more efficient and faster. Further, handles `None` in sequence length and `raw_rnn` style RNN classes does not work with a list of Tensors.

Implemented handling of `list` as input, so it does not break backwards compatibility.

### CODE REVIEW

**Comment by alrojo** [comment]
@ebrevdo Made some updates to allowing mix of list (current style) and tensors (new style). Further correct indentation and raises for wrong dimensions on logits, targets and weights.

**Comment by ebrevdo** [comment]
Lukasz will have to comment, but I believe we're starting a deprecation process for `seq2seq` which is monolithic and hard to maintain.  New code will sit somewhere in contrib.

**Comment by ebrevdo** [comment]
(it's unfortunate that you are sending all these PRs now, after we've just decided to start deprecating; sorry for the bad timing)

**Comment by lukaszkaiser** [comment]
Oh, while static seq2seq will be deprecated, the sequence_loss functions will stay (we might just move them to another module). I'm therefore reopening this. alrojo -- maybe instead of improving the seq2seq module you could put these functions in a new module, maybe tf.contrib.seq2seq? That one could be fully dynamic and ultimately (but before TF 1.0) replace seq2seq. Does this sound ok? Great thanks for your CLs, they help a lot!

**Comment by alrojo** [comment]
No problem, TensorFlow has helped me much. So I felt like giving back :)

I could work on a tf.contrib.seq2seq.py file. I also have a `dynamic_rnn_decoder_attention` ready for PR (based on the `raw_rnn` as well). I will start by making tests as advised in #4686 for the `dynamic_rnn_decoder` and I assume you would like kernel tests for this seq2seq loss function as well? After thoese two tasks I can make a seq2seq folder [here](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib) with similar structures to the other contrib folders.

I might have a few questions underway regarding setting up tests and correct folder structure. Would you like those questions to be on this thread (with an open PR), in the Issue section or another place? Thanks

**Comment by lukaszkaiser** [comment]
Yes, making a seq2seq folder in contrib sounds like the best way to go. Then you can have a separate file there for loss functions and maybe a separate one for decoders (in current seq2seq these things got somehow jumbled toegther). Great thanks for doing this, having a fully dynamic seq2seq will help a lot of people!

**Comment by lukaszkaiser** [comment]
P.S. As for questions, any thread is fine.

**Comment by ebrevdo** [comment]
Recommend focusing only on the dynamic case... Don't bother adding logic to
handle lists of inputs.

**Comment by vrv** [comment]
Any updates on this?  I'll close if this is no longer relevant.

**Comment by alrojo** [comment]
Yes, this will be a part of the loss function for the new dynamic decoder.

I will update this as a part of the new tf.contrib.seq2seq section once we have finished the [dynamic_decoder](https://github.com/tensorflow/tensorflow/pull/4686) PR.

**Comment by danmane** [comment]
Any update on this? Looks like maybe we should close it?

**Comment by lukaszkaiser** [comment]
Now alrojo is actively working on another related PR, I'm not sure we should close it yet.

**Comment by alrojo** [comment]
@danmane @lukaszkaiser I was waiting for #4686 to avoid merge conflicts. I will update this PR tomorrow. Thanks for your patience.

**Comment by alrojo** [comment]
@ebrevdo @lukaszkaiser the code has now been updated accordingly to the recent `contrib.seq2seq` commits. The code compiles without issues on my own jenkins replica for the Linux CPU test.

Please review, thanks!

Also, let me know if you're at NIPS and want to meet :)

**Comment by aleSuglia** [comment]
Hi to all,

I have seen the implementation of the seq2seq loss and it seems that it doesn't ignore possible padding values that can be used during training to make sequences of the same length. However, these values should not be considered in the loss function.. Am I wrong?

Thank you for your great work @alrojo. I hope that this pull request will be merged as soon as possible :)

Best regards,
Alessandro

**Comment by alrojo** [comment]
@aleSuglia With padding do you mean masking? The function supports masking.

**Comment by aleSuglia** [comment]
Yes I mean masking. I've read again your code and I think that you apply masking here: https://github.com/alrojo/tensorflow/blob/a6eafe994bd4c3fd66b4d0f79083279f72d358d1/tensorflow/contrib/seq2seq/python/ops/loss.py#L76

So the correct way to ignore the error in the prediction for padding values is done by setting to zero the elements of *weights* corresponding to padding values. Is it right?

Thank you for your help!

**Comment by alrojo** [comment]
@aleSuglia yes, that is correct. I have added a more detailed description to avoid confusion.

**Comment by alrojo** [comment]
@martinwicke Any news?

**Comment by lukaszkaiser** [comment]
Hi Alex. Is this code ready for review? What is missing? I think it got lost between the PRs. Let us know what needs to be done please!

**Comment by alrojo** [comment]
Hi @lukaszkaiser , yes, the code is finalized and ready for review.

File: tensorflow/contrib/seq2seq/python/kernel_tests/loss_test.py [hunk 1]
```diff
@@ -20,14 +20,58 @@
...
 import tensorflow as tf
 
-
 class LossTest(tf.test.TestCase):
 
-  def testLoss(self):
-    pass
+  def testSequenceLoss(self):
+    with self.test_session() as sess:
+      with tf.variable_scope("root", initializer=
```
Comment (lukaszkaiser) [inline tensorflow/contrib/seq2seq/python/kernel_tests/loss_test.py]: Please, do not separate lines on "=" like here (i.e., move the whole "initializer=" down).

**Review by lukaszkaiser — APPROVED** [review]
Thanks! One small comment, looks good to me otherwise.

**Comment by alrojo** [comment]
@lukaszkaiser please review again.

**Comment by lukaszkaiser** [comment]
@tensorflow-jenkins test this please

**Comment by lukaszkaiser** [comment]
It looks like //tensorflow/contrib/seq2seq:loss_test  is failing, could you take a look?

**Comment by alrojo** [comment]
Of course, I will try and have it ready for monday. :)

**Comment by alrojo** [comment]
Please try and test again

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Oh, `seq2seq_loss` was my faulty merge. Fixing.

**Comment by drpngx** [comment]
Actually, it was in this CL. Fixed.

Jenkins, test this please.

**Comment by drpngx** [comment]
Jenkins, test this please.
